com.unity.render-pipelines.universal/Runtime/LightCookieManager.cs

```diff
@@ -70,10 +73,121 @@ public static Settings GetDefault()
+
+                // For Recursion
+                if (start < end)
+                {
+                    int pivot = Partition<T>(data, start, end, compare);
+
+                    if (pivot >= 1)
+                        QuickSort<T>(data, start, pivot, compare);
+
+                    if (pivot + 1 < end)
+                        QuickSort<T>(data, pivot + 1, end, compare);
+                }
+            }
+
+            static T Median3Pivot<T>(T[] data, int start, int pivot, int end, Func<T, T, int> compare)
+            {
+                void Swap(int a, int b)
+                {
+                    var tmp = data[a];
+                    data[a] = data[b];
+                    data[b] = tmp;
+                }
+
+                if (compare(data[end], data[start]) < 0) Swap(start, end);
+                if (compare(data[pivot], data[start]) < 0) Swap(start, pivot);
+                if (compare(data[end], data[pivot]) < 0) Swap(pivot, end);
+                return data[pivot];
+            }
+
+            static int Partition<T>(T[] data, int start, int end, Func<T, T, int> compare)
+            {
+                int diff = end - start;
+                int pivot = start + diff / 2;
+
+                //var pivotValue = data[pivot];
+                var pivotValue = Median3Pivot(data, start, pivot, end, compare);
+
+                while (true)
+                {
+                    while (compare(data[start], pivotValue) < 0) ++start;
+                    while (compare(data[end], pivotValue) > 0) --end;
+
+                    if (start >= end)
+                    {
+                        return end;
+                    }
+
+                    var tmp = data[start];
+                    data[start++] = data[end];
+                    data[end--] = tmp;
+                }
+            }
+
+            // A non-allocating predicated sub-array insertion sort.
+            // Stable
+            static public void InsertionSort<T>(T[] data, int start, int end, Func<T, T, int> compare)
+            {
+                Assertions.Assert.IsTrue((uint)start < data.Length);
+                Assertions.Assert.IsTrue((uint)end < data.Length);
+
+                for (int i = start + 1; i < end + 1; i++)
+                {
+                    var iData = data[i];
+                    int j = i - 1;
+                    while (j >= 0 && compare(iData, data[j]) < 0)
+                    {
+                        data[j + 1] = data[j];
+                        j--;
+                    }
+                    data[j + 1] = iData;
+                }
+            }
+        }
+
         private struct LightCookieMapping
         {
             public ushort visibleLightIndex; // Index into visible light (src)
             public ushort lightBufferIndex;  // Index into light shader data buffer (dst)
+            public Light legacyLight; // Cached legacy light for the visibleLightIndex to avoid multiple copies on get from native array
```

Comment:
Minor comment; what do you mean with Legacy?

Reply:
It's the built-in light which doesn't fully describe a light in URP or HDRP and some of the fields aren't used.
I've seen it called "legacyLight" in a few places, there's no other reason.
I'll rename it to be "light".